snuba/state/rate_limit.py

```diff
@@ -119,24 +130,43 @@ def rate_limit(
                                   ^
                                  now
     """
-
-    bucket = "{}{}".format(state.ratelimit_prefix, rate_limit_params.bucket)
-    query_id = str(uuid.uuid4())
-
-    now = time.time()
-    bypass_rate_limit, rate_history_s = state.get_configs(
-        [("bypass_rate_limit", 0), ("rate_history_sec", 3600)]
-        #                               ^ number of seconds the timestamps are kept
+    bypass_rate_limit, rate_history_s, rate_limit_shard_factor = state.get_configs(
+        [
+            # bool (0/1) flag to disable rate limits altogether
+            ("bypass_rate_limit", 0),
+            # number of seconds the timestamps are kept
+            ("rate_history_sec", 3600),
+            # number of shards that each redis set is supposed to have.
+            # increasing this value multiplies the number of redis keys by that
+            # factor, and (on average) reduces the size of each redis set
+            ("rate_limit_shard_factor", 1),
+        ]
     )
     assert isinstance(rate_history_s, (int, float))
+    assert isinstance(rate_limit_shard_factor, int)
+    assert rate_limit_shard_factor > 0
 
     if bypass_rate_limit == 1:
         yield None
         return
 
+    now = time.time()
+
+    query_id_uuid = uuid.uuid4()
+    query_id = str(query_id_uuid)
+
+    bucket_shard = int(query_id_uuid) % rate_limit_shard_factor
```

Comment:
How does this work when we increase the factor? Won't we allocate queries to the wrong shards?

Reply:
@evanh I don't think so, because there's no consistency required across calls to this function. query_id itself is rolled at the start of this function, and as long as we keep using the same shard for the lifetime of the context manager, we should be good.

If we _decrease_ the number of shards, it may mean that we temporarily undercount rate limits as we stop zcounting sets that still have relevant data in them. This is a temporary effect that should fix itself within a couple minutes.

If we increase number øf shards, I think there should not be a problem because it just means we will read additional redis keys, but keep readinig the old ones.

I suggest we'll set this option once (or tweak it as little as possible) and just accept the dataloss.

If that's not acceptable we can also have two settings: how many shards to write, and how many shards to read. This would allow us to decrease shard count without dataloss.